include/engine/guidance/route_step.hpp

```diff
@@ -71,55 +71,117 @@ struct RouteStep
     // indices into the locations array stored the LegGeometry
     std::size_t geometry_begin;
     std::size_t geometry_end;
-    std::vector<Intersection> intersections;
+    std::vector<IntermediateIntersection> intersections;
 
-    LaneID numLanesToTheRight() const
+    // remove all information from the route step, marking it as invalid (used to indicate empty
+    // steps to be removed).
+    void Invalidate()
+    {
+        name_id = EMPTY_NAMEID;
+        name.clear();
+        ref.clear();
+        pronunciation.clear();
+        destinations.clear();
+        rotary_name.clear();
+        rotary_pronunciation.clear();
+        duration = 0;
+        distance = 0;
+        mode = TRAVEL_MODE_INACCESSIBLE;
+        maneuver = getInvalidStepManeuver();
+        geometry_begin = 0;
+        geometry_end = 0;
+        intersections.clear();
+        intersections.push_back(getInvalidIntersection());
+    }
+
+    // Elongate by another step in front
+    RouteStep &MakePreceededBy(const RouteStep &preceeding_step)
+    {
+        BOOST_ASSERT(preceeding_step.geometry_end == geometry_begin + 1);
+        BOOST_ASSERT(mode == preceeding_step.mode);
+        duration += preceeding_step.duration;
+        distance += preceeding_step.distance;
+
+        // when elongating in the front, we really don't want to introduce waypoint errors
+        BOOST_ASSERT(maneuver.waypoint_type == WaypointType::None &&
+                     preceeding_step.maneuver.waypoint_type == WaypointType::None);
+
+        geometry_begin = preceeding_step.geometry_begin;
+        intersections.insert(intersections.begin(),
+                             preceeding_step.intersections.begin(),
+                             preceeding_step.intersections.end());
+
+        return *this;
+    }
+
+    // Elongate by another step in back
+    RouteStep &ElongateBy(const RouteStep &following_step)
+    {
+        BOOST_ASSERT(geometry_end == following_step.geometry_begin + 1);
+        BOOST_ASSERT(mode == following_step.mode);
+        duration += following_step.duration;
+        distance += following_step.distance;
+
+        geometry_end = following_step.geometry_end;
+        intersections.insert(intersections.end(),
+                             following_step.intersections.begin(),
+                             following_step.intersections.end());
+
+        return *this;
+    }
+
+    // Elongate without prior knowledge of in front, or in back.
+    RouteStep &MergeWith(const RouteStep &by_step)
```

Comment:
If we have this functionality, why do we even expose the front / back functions above? Can't we always hide them from the user, preventing misuse?